### PR TITLE
Improved std.typecons.Tuple documentation.

### PR DESCRIPTION
**Pull request by MetaLang**
- Add examples
- Move old examples to documented unit tests
- Document certain behaviours (such as how tuples are compared)

### CODE REVIEW

**Comment by quickfur** [comment]
Is it possible to make most (if not all) of the code examples ddoc'd unittests? Those are preferred because they are verified to be compilable/runnable by the test suite.

**Comment by MetaLang** [comment]
I don't know. Can you add documented tests for member functions?

**Comment by quickfur** [comment]
Of course.
````
///
struct S
{
    ///
    void myMethod() { ... }
    ///
    unittest { ... }
}
````

**Comment by MetaLang** [comment]
Huh, I didn't know that was a thing. I'll change them to unittests then.

**Comment by MetaLang** [comment]
Done and commits squashed.

**Comment by quickfur** [comment]
Autotester failures, please fix.

File: std/typecons.d [hunk 1]
```diff
@@ -518,6 +472,23 @@ template Tuple(Specs...)
             // This is mostly to make t[n] work.
             alias _Tuple_super this;
         }
+        ///
+        unittest
+        {
+            auto t = tuple(1, " hello ", 2.3);
+            writeln(t);        // Tuple!(int, string, double)(1, " hello ", 2.3)
+            writeln(t.expand); // 1 hello 2.3
```
Comment (quickfur) [inline std/typecons.d]: The `writeln`s probably need `import std.stdio;`. However, being a unittest, it's probably better to use asserts instead. For example:
````
import std.format : format;
assert(format("%s", t) == "Tuple!(int, string, double)(1, " hello ", 2.3)");
````
and so on.

File: std/typecons.d [hunk 2]
```diff
@@ -518,6 +472,23 @@ template Tuple(Specs...)
...
+            
+            void takeSeveralTypes(int n, string s, double d) 
+            {
+                //Prints "4 test 5.1"
+                writeln(n, " ", s, " ", d);
```
Comment (quickfur) [inline std/typecons.d]: In a unittest, you can use `std.format.format` to verify that the output is actually what the comment claims it is.

File: std/typecons.d [hunk 3]
```diff
@@ -532,15 +503,16 @@ template Tuple(Specs...)
                 field[] = values[];
             }
         }
+        ///
+        unittest
+        {
+            alias ISD = Tuple!(int, string, double);
+            auto tup = ISD(1, "test", 3.2);
+            writeln(tup); //Prints Tuple!(int, string, double)(1, "test", 3.2)
```
Comment (quickfur) [inline std/typecons.d]: ditto

**Comment by MetaLang** [comment]
I guess this is why you should run the Phobos tests after adding documented unit tests. Didn't somebody make a change to the makefile recently to test a single module? I'm positive I saw that pull request but it doesn't look like it's in there.

**Comment by MetaLang** [comment]
All green.

**Comment by quickfur** [comment]
I don't bother with makefiles when I only need to test a single module. It's much faster and pain-free to run:
````
dmd -main -unittest std/algorithm/comparison.d -of/tmp/test && /tmp/test
````
which compiles a single module in unittest mode and runs it. This is on Linux, of course, I've no idea if this is possible or reliable on windows.

File: std/typecons.d [hunk 4]
```diff
@@ -698,9 +732,45 @@ template Tuple(Specs...)
         }
     }
 }
+///
```
Comment (quickfur) [inline std/typecons.d]: Nitpick: please insert blank line before the ddoc marker for each unittest block.

**Comment by quickfur** [comment]
Other than that, LGTM. This is a great improvement to the docs. Thanks!!

**Comment by JakobOvrum** [comment]
Nice. LGTM

**Comment by MetaLang** [comment]
Fixed the nits, ready to merge.

**Comment by quickfur** [comment]
LGTM. Thanks!

**Comment by quickfur** [comment]
Auto-merge toggled on

**Pull request merged**